ExitIntegrationTest.java

```diff
@@ -0,0 +1,176 @@
+import bcccp.tickets.season.UsageRecord;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import org.junit.Before;
+import org.junit.FixMethodOrder;
+import org.junit.Test;
+import org.junit.runners.MethodSorters;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+/**
+ *
+ * @author James Nocevski
+ */
+@FixMethodOrder(MethodSorters.NAME_ASCENDING)
+public class ExitIntegrationTest {
+    
+    IGate testGate;
+    CarSensor testIS;
+    CarSensor testOS; 
+    ExitUI testUI;
+    Carpark testCarpark;
+    AdhocTicket  testAdhoc;
+    SeasonTicket testSeason;
+    IAdhocTicketDAO mockAdhocDAO;
+    ISeasonTicketDAO mockSeasonDAO;
+    ExitController testExit;
+    String seasonTicketId, cParkName;
+    UsageRecord mockRecord;
+    int cParkCap;
+    
+    @Before
+    public void setUp() {
+        cParkName = "test name";
+        cParkCap = 5;
+        mockAdhocDAO = mock(AdhocTicketDAO.class);
+        mockSeasonDAO = mock(SeasonTicketDAO.class);
+        mockRecord = mock(UsageRecord.class);
+        testCarpark = new Carpark(cParkName, cParkCap, mockAdhocDAO, mockSeasonDAO);
+        testGate = new Gate(0, 0);
+        testIS = new CarSensor("test inside", 100, 100);
+        testOS = new CarSensor("test outside", 200, 200);
+        testUI = mock(ExitUI.class);
+        testAdhoc = new AdhocTicket(cParkName, 1, "Atest");
+        testSeason = new SeasonTicket("test", cParkName, 1234, 4321);
+        testExit = new ExitController(testCarpark, testGate, testIS, testOS, testUI);
+    }
+    
+    @Test
+    public void initialisationTest() {
+        assertTrue(testCarpark instanceof Carpark);
+        assertTrue(testIS instanceof CarSensor);
+        assertTrue(testOS instanceof CarSensor);
+        assertTrue(testExit instanceof ExitController);
+    }
+    
+    @Test
+    public void testNormalFlowSeason() {
+        assertEquals(testExit.getState(), "IDLE");  //checks initial state
+        testIS.setCarDetected(true);
+        testExit.carEventDetected(testIS.getId(), true);
+        
+        assertEquals(testExit.getState(), "WAITING");
+        verify(testUI).display("Insert Ticket");
+        
+        when(mockSeasonDAO.findTicketById(testSeason.getId())).thenReturn(testSeason);
+        testSeason.recordUsage(mockRecord);
+        testExit.ticketInserted(testSeason.getId());
+        assertEquals(testExit.getState(), "PROCESSED");
+        
+        testExit.ticketTaken();
+        verify(testUI).display("Take Processed Ticket");
+        assertEquals(testExit.getState(), "TAKEN");     
+        
+        testOS.setCarDetected(true);
+        testExit.carEventDetected(testOS.getId(), true);
+        assertEquals(testExit.getState(), "EXITING");
+        
+        testIS.setCarDetected(false);
+        testExit.carEventDetected(testIS.getId(), false);
+        assertEquals(testExit.getState(), "EXITED");

```

Comment:
again, you need to finish testing the use case
points 9, 12, 13 & 14 in the use case description have not been proven

9.       System raises entry barrier.
11.       System detects that car has exited
12.       System lowers exit barrier
13.       System records exit time for ticket
14.       System increments available spaces

Reply:
Fixed.

Reply:
Again these two appear not to have been checked:
12. System records exit time for ticket
13. System increments available spaces

Reply:
Fixed.